syntax = "proto2";

package rv.protos;

import "google/protobuf/struct.proto";
import "rastervision/protos/raster_transformer.proto";
import "rastervision/protos/vector_source.proto";

message RasterSourceConfig {
    message GeoTiffFiles {
        repeated string uris = 1;
        optional float x_shift_meters = 2;
        optional float y_shift_meters = 3;
    }

    message ImageFile {
        required string uri = 1;
    }

    message RasterioSource {
        repeated string uris = 1;
        optional float x_shift_meters = 2;
        optional float y_shift_meters = 3;
    }

    message ZXYRasterSource {
        required string tile_schema = 1;
        required int32 zoom = 2;
        repeated double bounds = 3;
        optional float x_shift_meters = 4;
        optional float y_shift_meters = 5;
    }

    // Used to read a VectorSource as a raster useful for semantic segmentation.
    message RasterizedSource {
        message RasterizerOptions {
            // The class_id to use for background pixels that don't overlap with any
            // shapes in the GeoJSON file.
            required int32 background_class_id = 2;

            // Here for backward compatibility. Now this is configured as part of the
            // VectorSource.
            // Number of pixels to add to each side of line when rasterized.
            optional int32 line_buffer = 3 [default=15];

            // Whether to rasterize all touched pixels
            optional bool all_touched = 4 [default=false];
        }
        required VectorSourceConfig vector_source = 1;
        required RasterizerOptions rasterizer_options = 2;
    }

    // Needed to be backwards compatible.
    message GeoJSONFile {
        message RasterizerOptions {
            // The class_id to use for background pixels that don't overlap with any
            // shapes in the GeoJSON file.
            required int32 background_class_id = 2;

            // Number of pixels to add to each side of line when rasterized.
            optional int32 line_buffer = 3 [default=15];
        }
        required string uri = 1;
        required RasterizerOptions rasterizer_options = 2;
    }

    required string source_type = 1;
    repeated RasterTransformerConfig transformers = 2;
    // The channel indices and order to use when extracting chips from a
    // RasterSource. For instance, this should be an array of length 3 to
    // generate RGB chips.
    repeated int32 channel_order = 3;

    oneof raster_source_config {
        GeoTiffFiles geotiff_files = 4;
        ImageFile image_file = 5;
        GeoJSONFile geojson_file = 6;
        google.protobuf.Struct custom_config = 7;
        RasterizedSource rasterized_source = 8;
        RasterioSource rasterio_source = 9;
        ZXYRasterSource zxy_raster_source = 10;
    }
}
